File: harness.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

// SEAM: a better system prompt is the cheapest improvement you can make
// (Suggestion 2).
const systemPrompt = "You are BadClaude, a helpful assistant."

// Harness is the heart of BadClaude: read a line, send it to the model,
// print the reply. That's it. Every improvement you make plugs in here or
// replaces a piece this type uses.
type Harness struct {
	llm    *LLMClient
	memory Memory
	log    *InteractionLog

	// SEAM: register your tools here and teach the model to ask for them
	// (Suggestions 3 and 4).
	tools []Tool
}

// NewHarness puts the pieces together.
func NewHarness(llm *LLMClient, memory Memory, log *InteractionLog) *Harness {
	return &Harness{llm: llm, memory: memory, log: log, tools: []Tool{}}
}

// Run is the chat loop. It returns when you type /quit or input ends.
func (h *Harness) Run() {
	fmt.Println("BadClaude is listening. Type /quit to exit.")
	fmt.Println()

	reader := bufio.NewReader(os.Stdin)
	for {
		fmt.Print("you> ")
		line, err := reader.ReadString('\n')
		if err != nil && line == "" {
			break // end of input
		}
		input := strings.TrimSpace(line)
		if input == "" {
			continue
		}
		if input == "/quit" {
			break
		}

		h.log.Record("user", input)
		userMessage := UserMessage(input)

		// Build the request: system prompt, whatever the memory recalls,
		// then the new user message. With NoMemory, the model sees only
		// the newest message -- try asking it about your previous one!
		request := []Message{SystemMessage(systemPrompt)}
		for _, remembered := range h.memory.Recall() {
			request = append(request, remembered)
		}
		request = append(request, userMessage)

		reply, err := h.llm.Chat(request)
		if err != nil {
			problem := err.Error()
			fmt.Println("[error] " + problem)
			h.log.Record("error", problem)
			continue
		}

		h.memory.Add(userMessage)
		h.memory.Add(AssistantMessage(reply))
		h.log.Record("assistant", reply)

		// SEAM: this is where an agent loop would check the reply for a
		// tool request, run the tool, and go back to the model with the
		// result instead of printing straight away (Suggestion 4).
		fmt.Println()
		fmt.Println("badclaude> " + reply)
		fmt.Println()
	}

	h.log.Close()
	fmt.Println("Bye!")
}

File: src/harness.rs
use crate::interaction_log::InteractionLog;
use crate::llm_client::LlmClient;
use crate::memory::Memory;
use crate::message::Message;
use crate::tool::Tool;
use std::io;
use std::io::Write;

// SEAM: a better system prompt is the cheapest improvement you can make
// (Suggestion 2).
const SYSTEM_PROMPT: &str = "You are BadClaude, a helpful assistant.";

/// The heart of BadClaude: read a line, send it to the model, print the reply.
/// That's it. Every improvement you make plugs in here or replaces a piece
/// this struct uses.
pub struct Harness {
    llm: LlmClient,
    memory: Box<dyn Memory>,
    log: InteractionLog,

    // SEAM: register your tools here and teach the model to ask for them
    // (Suggestions 3 and 4).
    //
    // (Rust warns about anything that is never used. Nothing uses this list
    // yet -- that's your job -- so we tell the compiler that's on purpose.
    // Delete the #[allow] line once your harness uses the tools.)
    #[allow(dead_code)]
    tools: Vec<Box<dyn Tool>>,
}

impl Harness {
    pub fn new(llm: LlmClient, memory: Box<dyn Memory>, log: InteractionLog) -> Harness {
        Harness {
            llm,
            memory,
            log,
            tools: Vec::new(),
        }
    }

    pub fn run(&mut self) {
        println!("BadClaude is listening. Type /quit to exit.");
        println!();

        let stdin = io::stdin();
        loop {
            print!("you> ");
            // print! doesn't show anything until the line ends, unless we
            // flush it ourselves.
            let _ = io::stdout().flush();

            let mut line = String::new();
            match stdin.read_line(&mut line) {
                Ok(0) => break,  // end of input
                Err(_) => break, // can't read input
                Ok(_) => {}
            }
            let input = line.trim().to_string();
            if input.is_empty() {
                continue;
            }
            if input == "/quit" {
                break;
            }

            self.log.record("user", &input);
            let user_message = Message::user(&input);

            // Build the request: system prompt, whatever the memory recalls,
            // then the new user message. With NoMemory, the model sees only
            // the newest message -- try asking it about your previous one!
            let mut request = Vec::new();
            request.push(Message::system(SYSTEM_PROMPT));
            for message in self.memory.recall() {
                request.push(message);
            }
            request.push(user_message.clone());

            let reply = match self.llm.chat(&request) {
                Ok(reply) => reply,
                Err(problem) => {
                    println!("[error] {}", problem);
                    self.log.record("error", &problem);
                    continue;
                }
            };

            self.memory.add(user_message);
            self.memory.add(Message::assistant(&reply));
            self.log.record("assistant", &reply);

            // SEAM: this is where an agent loop would check the reply for a
            // tool request, run the tool, and go back to the model with the
            // result instead of printing straight away (Suggestion 4).
            println!();
            println!("badclaude> {}", reply);
            println!();
        }

        self.log.close();
        println!("Bye!");
    }
}
